### ISSUE TITLE
Divergence issue with L2 loss (you have to adjust the learning rate accordingly)

### ISSUE THREAD

**Issue by Jongchan** [cvpa3/dl-practice#4]
It seems the output ground truth label size is different when using L2 loss (nn.MSECriterion), 
and it seems inappropriate to use log softmax as the final output of the network.

Thus, I changed the final output of the network as a normal 10 neurons, and the ground truth label as one-hot vector with length 10.

During training, I found the loss diverges quickly to **'nan'**, and I thought I was doing something wrong with the network/loss definition.

It turned out that **the default learning rate (1e-2,1e-2)** was **too high** for this regression task.
Learning rate of **1e-4** is the right number for me, for now.

**Comment by barduris** [comment]
I was running into the same issue, and adjusting the learning rate fixed it.

I also agree with you on the use of log softmax. It seems strange to use it with the L2 loss. I personally replaced it with a tanh for that particular case.

**Comment by Jongchan** [comment]
@barduris
tanh.. That would have been better :)

**Comment by mcahny** [comment]
I have experimented with many combinations of different activation functions and learning rates, 

and the combi of HardTanh, and learning rate of 1e-3 worked best for me. Labels used look like {-1,-1,1,-1,-1,-1,-1,-1,-1,-1} in this case. 

As @barduris suggested, Tanh with normal one-hot label({0,0,...,1,..,0}) also worked, with a little bit lower accuracy. 

With Sigmoid, PReLU, and without any activation, I could not find suitable learning rates, and they don't seem to learn very well showing the accuracy below 30%

**Comment by mcahny** [comment]
I couldn't find suitable learning rate when optimizing L2 loss without any activation,
bus as @barduris suggested, Tanh at the end of the network works fine for me.
